File: DsAndAlgo/Python/src/Practice/src/linkedlist_in_binary_tree.py
from typing import Optional

from src.Practice.ds.list_node import ListNode
from src.Practice.ds.tree_node import TreeNode


class LinkedListInBinaryTree:

  @staticmethod
  def dfs_vist(head: Optional[ListNode], root: Optional[TreeNode]) -> bool:

        """
        Check if the linked list is equal to the path in the binary tree.

        :param head: The head of the linked list.
        :param root: The current node in the binary tree.
        :return: True if the linked list is equal to the path in the binary tree, False otherwise.
        """
        if not head:
            return True
        if not root:
            return False
        if head.val != root.val:
            return False
        return (LinkedListInBinaryTree.dfs_vist(head.next, root.left) or
                LinkedListInBinaryTree.dfs_vist(head.next, root.right))

  @staticmethod
  def is_sub_path(head: Optional[ListNode], root: Optional[TreeNode]) -> bool:

        """
        Check if a linked list is a subpath in a binary tree.

        :param head: The head of the linked list.
        :param root: The root of the binary tree.
        :return: True if the linked list is a subpath in the binary tree, False otherwise.
        """
        if not head:
            return True
        if not root:
            return False
        if LinkedListInBinaryTree.dfs_vist(head, root):
            return True
        return (LinkedListInBinaryTree.is_sub_path(head, root.left) or
                LinkedListInBinaryTree.is_sub_path(head, root.right))

File: DsAndAlgo/Java/Coding/src/main/java/org/practice/LinkedListInBinaryTree.java
package org.practice;

import com.sun.source.tree.Tree;

public class LinkedListInBinaryTree {

    public static boolean dfsVisit(ListNode listNode, TreeNode treeNode) {
        if (listNode == null) {
            return true; // Reached the end of the linked list
        }
        if (treeNode == null) {
            return false; // Reached a leaf node in the tree without matching the list
        }
        if (listNode.val != treeNode.val) {
            return false; // Values do not match
        }
        // Continue searching in both left and right subtrees
        return dfsVisit(listNode.next, treeNode.left) || dfsVisit(listNode.next, treeNode.right);

    }
    public static boolean isSubPath(ListNode head, TreeNode root) {
        if (head == null) return  true;
        if (root == null) return false;
        // Start DFS from the root of the tree
        if (dfsVisit(head, root)) {
            return true; // Found a matching path
        }
        return  isSubPath(head, root.left) || isSubPath(head, root.right);
    }

}
